include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,222 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+
+#pragma once
+#ifndef __WIL_CPPWINRT_AUTHORING_INCLUDED
+#define __WIL_CPPWINRT_AUTHORING_INCLUDED
+
+namespace wil
+{
+    template <typename T>
+    struct read_only_property
+    {
+        read_only_property(T value) : m_value(value) { }
+
+        operator T() const noexcept { return m_value; }
+        const T& operator()() const noexcept { return m_value; }
+
+        read_only_property(const read_only_property& other) noexcept : m_value(other.m_value) { }
+        read_only_property(read_only_property&& other) noexcept : m_value(std::move(other.m_value)) { }
+
+        template<typename... TArgs>
+        read_only_property(TArgs&&... args) : m_value(std::forward<TArgs>(args)...) { }
+
+    protected:
+        T m_value;
+    };
+
+    template <typename T>
+    struct read_write_property : read_only_property<T>

```

Comment:
I'll move the discussion here since the original comment got resolved and the file name changed. The original comment was regarding the fact that this type does nothing to facilitate synchronization.

I understand that people can still manually write properties like they do today. My concern is more that it's too easy to see this - or perhaps read code that uses it - and assume that it's the "Microsoft recommended" way to implement properties. I could _maybe_ understand if this is intended for something like a class whose properties are only going to be read/written by XAML, though in that case (and the general case) I'd think that this _at least_ requires a different name. E.g. something like `single_threaded_read_write_property`.

Reply:
renamed to:
- `single_threaded_ro_property`
- `single_threaded_rw_property`
- `single_threaded_notifying_property`

a little verbose IMO, but i can live with that and i'll probably alias those types in my usage to something shorter.